nautobot/extras/models/groups.py

```diff
@@ -0,0 +1,313 @@
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django.urls import reverse
+
+from nautobot.core.models.generics import OrganizationalModel
+from nautobot.extras.groups import dynamicgroup_map_factory
+from nautobot.extras.utils import extras_features
+from nautobot.utilities.querysets import RestrictedQuerySet
+
+
+logger = logging.getLogger(__name__)
+
+
+class DynamicGroupQuerySet(RestrictedQuerySet):
+    """Queryset for `DynamicGroup` objects."""
+
+    def get_for_object(self, obj):
+        """
+        Return all `DynamicGroup` assigned to the given object.
+        """
+        if not isinstance(obj, models.Model):
+            raise TypeError(f"{obj} is not an instance of Django Model class")
+
+        # Get dynamic groups for this content_type using the discrete content_type fields to
+        # optimize the query.
+        # TODO(jathan): 1 query
+        eligible_groups = self.filter(
+            content_type__app_label=obj._meta.app_label, content_type__model=obj._meta.model_name
+        ).select_related("content_type")
+
+        # Filter down to matching groups
+        my_groups = []
+        # TODO(jathan: 3 queries per DynamicGroup instance
+        for dynamic_group in eligible_groups:
+            if obj.pk in dynamic_group.get_queryset(flat=True):
+                my_groups.append(dynamic_group.pk)
+
+        # TODO(jathan): 1 query
+        return self.filter(pk__in=my_groups)
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "custom_validators",
+    "dynamic_groups",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "webhooks",
+)
+class DynamicGroup(OrganizationalModel):
+    """Dynamic Group Model."""
+
+    name = models.CharField(max_length=100, unique=True, help_text="Dynamic Group name")
+    slug = models.SlugField(max_length=100, unique=True, help_text="Unique slug")
+    description = models.CharField(max_length=200, blank=True)
+    content_type = models.ForeignKey(
+        to=ContentType,
+        on_delete=models.CASCADE,
+        verbose_name="Object Type",
+        help_text="The type of object for this group.",
+    )
+    # TODO(jathan): Set editable=False that this field doesn't show up in forms.
+    # It's a complex field only modified by the DynamicGroupForm internals at
+    # this time. I am not happy with this pattern right now but due to the
+    # dynamism of the form by merging in the `FooFilterForm` fields, there's not
+    # currently an easy way to move this construction logic to the model.
+    filter = models.JSONField(
+        encoder=DjangoJSONEncoder,
+        editable=False,
+        default=dict,
+        blank=True,
+        null=True,
+        help_text="A JSON-encoded dictionary of filter parameters for group membership",
+    )
+
+    objects = DynamicGroupQuerySet.as_manager()
+
+    clone_fields = ["content_type", "filter"]
+
+    class Meta:
+        ordering = ["content_type", "name"]
```

Comment:
Does this work as expected for ordering, or should it be something like `["content_type__app_label", "content_type__model", "name"]` instead?